KAFKA-6573: Update brokerInfo in KafkaController on listener update

rajinisivaram opened the pull request:
Update KafkaController.brokerInfo when listeners are updated since this value is used to register the broker in ZooKeeper if there ZK session expires. Also added test to verify values in ZK after session expiry.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rajinisivaram committed:
KAFKA-6573: Update brokerInfo in KafkaController on listener update

rajinisivaram commented:
@hachikuji Can you review, please? Thank you!

hachikuji commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -599,13 +599,34 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+      assertTrue(s"Endpoint update not saved $endpoints", endpointMatcher(endpoints))
+    }
+
+    // Verify that endpoints have been updated in ZK fr all brokers
```
typo: "for"?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -751,7 +751,9 @@ class DynamicListenerConfig(server: KafkaServer) extends BrokerReconfigurable wi
 
-    server.zkClient.updateBrokerInfoInZk(server.createBrokerInfo)
+    val newBrokerInfo = server.createBrokerInfo
+    server.zkClient.updateBrokerInfoInZk(newBrokerInfo)
```
Would it be reasonable to move this into `KafkaController.updateBrokerInfo` since initial registration is already done from that class?

hachikuji commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -599,13 +599,34 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+
+    // Trigger session expiry and ensure that controller registers new advertised listener after expiry
+    val controllerEpoch = zkClient.getControllerEpoch
+    val controllerServer = servers(zkClient.getControllerId.getOrElse(throw new IllegalStateException("No controller")))
```
Does this test case actually depend on expiring the controller's session? I thought the registration logic is common on all brokers.

rajinisivaram replied:
The test doesn't depend on expiring the controller's session, it could be any broker's ZK session. I chose the controller because I thought it was worth checking that the controller epoch had changed as a result of the expiry (not strictly required for the test though).

hachikuji reviewed (commented):
Good catch. Left some minor comments.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thanks for the review. I have addressed the comments and left a note for one. Actually I didn't find this, Ismael had noticed that I had changed `brokerInfo` from `val` to a `var` but there was no update of the `var` (Thanks @ijuma).

hachikuji reviewed (approved):
LGTM. Merging to trunk and 1.1. cc @dguy
